keps/sig-auth/4317-pod-certificates/README.md

```diff
@@ -0,0 +1,1428 @@
+
+**Support human-in-the-loop approval for certificate requests**: This KEP
+assumes that PodCertificateRequest issuance is fully-automated, because this
+process blocks pod startup for pods that use PodCertificate projected volume
+sources.  PodCertificateRequests do not separate the approval and issuance
+steps, unlike CertificateSigningRequests.
+
+**Standardize on using the SPIFFE SVID format in core Kubernetes**: The
+certificates issued by `kubernetes.io/kube-apiserver-client-pod` can be made
+compatible with SPIFFE in the future, since they currently don't include any
+Subject Alternate Name entries.
+
+**Support identity delegation or impersonation**: We do not intend to support
+issuing certificates that workloads can use as intermediate CA certificates in
+order to issue further certificates, nor do we intend to support one privileged
+workload minting certificates on behalf of other workloads.
+
+## Design Details
+
+### PodCertificateRequest Resource
+
+PodCertificateRequest is a new API type in certificates.k8s.io/v1alpha1.  It's a
+stripped-down version of CertificateSigningRequest that is tailored directly to
+the pod certificate issuance use case.
+
+PodCertificateRequest is fundamentally a statement that "Pod X is requesting a
+certificate from Signer Y".  The signer is in complete control of the form of
+the issued certificate.  As such, PodCertificateRequest objects don't contain a
+typical X.509 CSR.  Instead, they contain:
+* A public key, PKIX-serialized.  This key must be an RSA key with modulus size
+  3072 or 4096, an ECDSA key with curve P256 or P384, or an ED25519 key.
+* A proof of possession of the corresponding private key.
+  * If the key is an RSA key, then the signature is over the ASCII bytes of the
+	  pod UID, using RSASSA-PKCS1-V1_5-SIGN from RSA PKCS #1 v1.5 (as implemented
+	  by the golang function crypto/rsa.SignPKCS1v15).
+  * If the key is an ECDSA key, then the signature is as described by [SEC 1,
+	  Version 2.0](https://www.secg.org/sec1-v2.pdf) (as implemented by the golang
+	  library function crypto/ecdsa.SignASN1)
+  * If the key is an ED25519 key, then the signature is as described by the
+    [ED25519 Specification](https://ed25519.cr.yp.to/) (as implemented by
+    the golang library crypto/ed25519.Sign).
+* Details of the requesting Pod's identity:
+  * Namespace (indicated by the namespace the PodCertificateRequest is created in).
+  * Pod name
+  * Pod UID
+  * Service account name
+  * Service account UID
+  * Node name
+  * Node UID
+* The signer name from which a certificate is being requested.
+* (In the status subobject): An issued certificate chain, stored as a single
+  PEM-formatted string.  If there is more than one issued certificate, the first
+  element of the chain is assumed to be the leaf certificate, followed by any
+  necessary intermediates in leaf-to-root order.
+* (In the status subobject): A timestamp at which time the requester that made
```

Comment:
You mention this at two points, let's discuss it overall here.

While in my perfect world, all certificates would be short-lived, I don't think we can be this prescriptive for third-party signers.

For Kubernetes built-in signers, we control the code, so we can make sure the certificates are short-lived.  Additional validation isn't needed for our use cases.

If a third party wants to come and build a PodCertificateRequest signer called `my-cool-signer.com/lets-encrypt` that issues 90-day certs from LetsEncrypt, I don't think we should stop them.  They can assess the revocation story for their certs (which WebPKI certs will have) and decide whether or not its a good idea.

That's not to say we shouldn't do something about revocation within the framework of the Kubernetes API.  But it's a complicated problem with many sub-optimal solutions, and I don't think we have the time to figure out a workable approach before freeze.

For right now, a third-party signer implementation can bring their own revocation story.  It could be as simple as having their signing controller export the CRL in a configmap, and then making sure their workloads mount the configmap at a well-known pod.  To use their third-party certs, they'll have to be using their own third-party library code (not client-go), so they can make sure they plumb the CRL.

Secondly, even if we do decide to limit the lifetime of certs that can be issued through PodCertificateRequests, it should probably be part of validation on kube-apiserver, rather than in kubelet.

Reply:
We discussed most of this offline but re: future revocation support - I think we could provide full support in client-go for an embedded CRL distributed via a mounted CTB.  I am not worried about it for alpha, but I do want to push towards a future where revocation is the norm and not the exception in our ecosystem.

Reply:
Agreed.  I've also updated the text to reflect our conversation.

Specifically:
* PodCertificateRequests and volume specs now carry `maxExpirationSeconds`.
* Kube-apiserver will restrict `maxExpirationSeconds` based on a per-signer configuration.
* Kube-apiserver will enforce that the signer actually issued a certificate that conforms to `maxExpirationSeconds`